ckanext/validation/utils.py

```diff
@@ -1,73 +1,304 @@
+from six import string_types
 
+import ckan.plugins.toolkit as tk
+import ckan.plugins as plugins
+from ckan.model import Session
+from ckan.common import asbool, config
+from ckan.lib.uploader import (ALLOWED_UPLOAD_TYPES, _get_underlying_file,
+                               get_resource_uploader, ResourceUpload)
+
+import ckanext.validation.settings as settings
+from ckanext.validation.interfaces import IDataValidation
+from ckanext.validation.jobs import run_validation_job, _get_site_user_api_key
+from ckanext.validation.validation_status_helper import ValidationStatusHelper, StatusTypes
 
 log = logging.getLogger(__name__)
 
 
-def get_update_mode_from_config():
-    if asbool(
-            config.get(u'ckanext.validation.run_on_update_sync', False)):
+def get_update_mode():
+    if asbool(config.get(u'ckanext.validation.run_on_update_sync', False)):
         return u'sync'
-    elif asbool(
-            config.get(u'ckanext.validation.run_on_update_async', True)):
+    elif asbool(config.get(u'ckanext.validation.run_on_update_async', True)):
         return u'async'
     else:
         return None
 
 
-def get_create_mode_from_config():
-    if asbool(
-            config.get(u'ckanext.validation.run_on_create_sync', False)):
+def get_create_mode():
+    if asbool(config.get(u'ckanext.validation.run_on_create_sync', False)):
         return u'sync'
-    elif asbool(
-            config.get(u'ckanext.validation.run_on_create_async', True)):
+    elif asbool(config.get(u'ckanext.validation.run_on_create_async', True)):
         return u'async'
     else:
         return None
 
 
-def get_local_upload_path(resource_id):
+def process_schema_fields(data_dict):
     u'''
-    Returns the local path to an uploaded file give an id
+    Normalize the different ways of providing the `schema` field
+
+    1. If `schema_upload` is provided and it's a valid file, the contents
+        are read into `schema`.
+    2. If `schema_url` is provided and looks like a valid URL, it's copied
+        to `schema`
+    3. If `schema_json` is provided, it's copied to `schema`.
 
-    Note: it does not check if the resource or file actually exists
+    All the 3 `schema_*` fields are removed from the data_dict.
+    Note that the data_dict still needs to pass validation
     '''
-    upload = ResourceUpload({u'url': u'foo'})
-    return upload.get_path(resource_id)
 
+    schema_upload = data_dict.pop(u'schema_upload', None)
+    schema_url = data_dict.pop(u'schema_url', None)
+    schema_json = data_dict.pop(u'schema_json', None)
 
-def delete_local_uploaded_file(resource_id):
-    u'''
-    Remove and uploaded file and its parent folders (if empty)
+    if isinstance(schema_upload, ALLOWED_UPLOAD_TYPES) \
+            and schema_upload.filename:
+        data_dict[u'schema'] = _get_underlying_file(schema_upload).read()
+    elif schema_url:
+        if (not isinstance(schema_url, string_types)
+                or not schema_url.lower()[:4] == u'http'):
```

Comment:
This isn't a good method to check for a URL. `urlparse` might do better.

Reply:
\+

Reply:
This will be fixed and pushed by Alex shortly.